### PR TITLE
ipinfo man page

### PR DESCRIPTION
**Pull request by Taimoor-12**
Includes the man page written in `roff` markup, parsing the man page to print formatted `--help` output on the terminal and handling the installation of the man page to the user's local machine.

### CODE REVIEW

**Comment by Taimoor-12** [comment]
@UmanShahzad

File: debian/files [hunk 1]
```diff
@@ -1 +1 @@
-ipinfo_3.2.0_source.buildinfo utils optional
+ipinfo_3.2.0_source.buildinfo utils optional
```
Comment (UmanShahzad) [inline debian/files:1]: Pls add back the newline

File: debian/ipinfo.manpages [hunk 2]
```diff
@@ -0,0 +1 @@
+ipinfo/ipinfo.1
```
Comment (UmanShahzad) [inline debian/ipinfo.manpages:1]: Add a newline

File: debian/rules [hunk 3]
```diff
@@ -1,3 +1,17 @@
...
 %:
 	PATH=${PATH}:/usr/lib/go-1.18/bin dh $@ --builddirectory=_build --buildsystem=golang --with=golang
+
+	
+override_dh_auto_build:
+	cp ipinfo/ipinfo.1 _build/src/github.com/ipinfo/cli/ipinfo
+	dh_auto_build
+	
+
+
```
Comment (UmanShahzad) [inline debian/rules:11]: Kill all the extra newlines

File: debian/rules [hunk 4]
```diff
@@ -1,3 +1,17 @@
...
+override_dh_auto_install:
+	dh_auto_install
+
```
Comment (UmanShahzad) [inline debian/rules:14]: Rm

File: scripts/build-archive-all.sh [hunk 5]
```diff
@@ -10,6 +10,7 @@ ROOT=$DIR/..
 CLI=$1
 VSN=$2
 
+
```
Comment (UmanShahzad) [inline scripts/build-archive-all.sh:13]: Rm

File: scripts/build-archive-all.sh [hunk 6]
```diff
@@ -44,4 +45,9 @@ cd ..
 rm -rf $ROOT/${CLI}/dist/usr
 mkdir -p $ROOT/${CLI}/dist/usr/local/bin
 cp $ROOT/build/${CLI}_${VSN}_linux_amd64 $ROOT/${CLI}/dist/usr/local/bin/${CLI}
+
+# Copy man page to the appropriate location
+mkdir -p $ROOT/${CLI}/dist/usr/share/man/man1
+cp $ROOT/ipinfo/ipinfo.1 $ROOT/${CLI}/dist/usr/share/man/man1/
```
Comment (UmanShahzad) [inline scripts/build-archive-all.sh:51]:
```suggestion
cp $ROOT/${CLI}/${CLI}.1 $ROOT/${CLI}/dist/usr/share/man/man1/
```

File: scripts/build-archive-all.sh [hunk 7]
```diff
@@ -44,4 +45,9 @@ cd ..
 rm -rf $ROOT/${CLI}/dist/usr
 mkdir -p $ROOT/${CLI}/dist/usr/local/bin
 cp $ROOT/build/${CLI}_${VSN}_linux_amd64 $ROOT/${CLI}/dist/usr/local/bin/${CLI}
+
+# Copy man page to the appropriate location
```
Comment (UmanShahzad) [inline scripts/build-archive-all.sh:49]:
```suggestion
```

**Review by UmanShahzad — COMMENTED** [review]
- Can you show how the new `-h` looks now?
- We've made our "source file" be the actual manpage format now basically, and it's actually not a very good format to write in, as compared to markdown for example.

I'm looking at this with fresh eyes and thinking it may not be worth the maintenance burden to maintain the docs like this. If we change things, even testing the CLI's `-h` will become cumbersome as we'd have to re-build and re-run each time.

What I recall the original motivation on this was, it was to allow `ipinfo`'s help to be similar as `git` - when you do `-h`, it gives a short help, and when you do `--help`, it opens up a reader like `less` and shows extremely detailed documentation with many examples and explanations.

Can you check how git does `--help` like that but keeps `-h` as a short desc? It's actually using a manual page for `--help` but how do we just 'open' the manpage with `--help` transparently like that?

**Comment by Taimoor-12** [comment]
This is how `-h` or `--help` for `ipinfo` looks like:
ipinfo -h


Currently, `git --help` also prints something on the terminal instead of opening the man page:
git --help


This is what `git -h` looks like:
git -h






Upon further research into how Git maintains documentation for its `git` command, it appears that the source file is a plain text file written in `AsciiDoc` format—a more human-readable option. This text file contains the same sections as the man page, including NAME, SYNOPSIS, DESCRIPTION, COMMANDS, OPTIONS, EXAMPLES, etc. This file serves as the source for both the man page (often converted from `AsciiDoc` format) and the `--help` output. In either case, we would need to custom-parse the source file to extract the desired sections for printing the `--help` output. Instead of parsing the potentially cumbersome man page format, we can parse this more human-readable file.

**Comment by UmanShahzad** [comment]
@Taimoor-12 let's try an approach that doesn't involve any actual `man` file, but acts like one.

The main benefit I personally find from a `man` page is that it's very detailed and searchable using commands you find in pagers like `less`.

So can we find a way where `-h` (short help) is the same as today, but `--help` (long help) ends up printing a different help string for the same command that's the "full detailed" one and that **runs in a pager**? The key part I don't know myself is how would a CLI open something in a pager. I've seen e.g. `psql` open up a pager when the results returned from making a query are too long, for example, so it must be possible.

**Comment by Taimoor-12** [comment]
@UmanShahzad If I understand correctly, we want to put all the information about the `ipinfo` command and its subcommands, including examples and options, into a simple text file. When `--help` is passed, we will pipe the contents of that file to `less`. For `-h`, we can continue using the current hard-coded string version. Is that accurate?

**Comment by UmanShahzad** [comment]
Even for `--help` it can be a hardcoded string inside Go (or embedded via Go embed), but just paged instead of printed, so we can freely write pages and pages of details and examples.

**Comment by Taimoor-12** [comment]
What kind of a structure should we follow for the hardcoded string that will be paged with `less` when `--help` is used?

**Comment by UmanShahzad** [comment]
@Taimoor-12 we can follow our own structure at this point - anything that just allows us to communicate a lot more examples than currently. It could be the same structure but with just more details, explaining all edge cases, etc.

**Comment by Taimoor-12** [comment]
The output for `--help` after the new commit is this:

less

**Review by UmanShahzad — COMMENTED** [review]
Oh we're just executing `less` quite literally. Is that how this is done in CLIs like psql or something to achieve the pager?

**Comment by Taimoor-12** [comment]
I think yes. I encountered a CLI called `pgcli` and went through their docs on the website. They mentioned that the default pager for displaying long result sets is `less`. In their config file, users can explicitly specify the pager value, and the specified pager would be executed to display result sets (or simply `less` if nothing is specified). That's what I understood

Through the use of flags and when `--help` is invoked, I directly executed `less` on the long string.

**Comment by Taimoor-12** [comment]
Yes the pager can be set through an env var called `PAGER` which is a common name to set the pager tool. We can read its value and use that pager for the string. If the user has set no `PAGER` value, we can then use `less` as a pager.

**Comment by UmanShahzad** [comment]
Alright let's incorporate that!

File: scripts/build-archive-all.sh [hunk 8]
```diff
@@ -44,4 +44,4 @@ cd ..
 rm -rf $ROOT/${CLI}/dist/usr
 mkdir -p $ROOT/${CLI}/dist/usr/local/bin
 cp $ROOT/build/${CLI}_${VSN}_linux_amd64 $ROOT/${CLI}/dist/usr/local/bin/${CLI}
-dpkg-deb -Zgzip --build ${ROOT}/${CLI}/dist build/${CLI}_${VSN}.deb
+dpkg-deb -Zgzip --build ${ROOT}/${CLI}/dist build/${CLI}_${VSN}.deb
```
Comment (UmanShahzad) [inline scripts/build-archive-all.sh:47]: Pls add back the newline

File: ipinfo/cmd_default.go [hunk 9]
```diff
@@ -90,7 +115,8 @@ func cmdDefault() (err error) {
 	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
 	pflag.BoolVar(&fNoCache, "nocache", false, "disable the cache.")
 	pflag.BoolVarP(&fVsn, "version", "v", false, "print binary release number.")
-	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.BoolVarP(&fHelp, "short-help", "h", false, "show help.")
+	pflag.BoolVar(&fPage, "help", false, "use pager for detailed help")
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:119]:
```suggestion
	pflag.BoolVar(&fPage, "help", false, "show detailed help")
```

File: ipinfo/cmd_default.go [hunk 10]
```diff
@@ -90,7 +115,8 @@ func cmdDefault() (err error) {
 	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
 	pflag.BoolVar(&fNoCache, "nocache", false, "disable the cache.")
 	pflag.BoolVarP(&fVsn, "version", "v", false, "print binary release number.")
-	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.BoolVarP(&fHelp, "short-help", "h", false, "show help.")
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:118]: Does it work if we remove the long-hand form? (i.e. `--short-help`, just make it empty string?)

File: ipinfo/cmd_default.go [hunk 11]
```diff
@@ -108,6 +134,12 @@ func cmdDefault() (err error) {
 		return nil
 	}
 
+	if fPage {
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:137]: Shouldn't reference 'page' like that - we can just call it `fHelpDetailed` to be more clear.

Function should be similarly renamed

File: ipinfo/cmd_default.go [hunk 12]
```diff
@@ -77,6 +81,27 @@ Options:
 `, progBase)
 }
 
+func pageHelp(detailedHelp string) error {
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:84]: If this is going to be a re-usable utility then let's move it into `lib/`.

File: ipinfo/cmd_default.go [hunk 13]
```diff
@@ -77,6 +81,27 @@ Options:
...
+	cmd := exec.Command(pagerCmd)
+
+	// Create an io.Reader from the detailed help string
+	reader := io.Reader(strings.NewReader(detailedHelp))
+	cmd.Stdin = reader
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	return cmd.Run()
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:102]: We shouldn't return this. Instead, run the command, and if there was a problem e.g. the command didn't exist, we could back up to calling `printHelp` (short help) as a safe fallback.

Please do test if that fallback works if your pager is having problems.

File: debian/rules [hunk 14]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/make -f
 %:
-	PATH=${PATH}:/usr/lib/go-1.18/bin dh $@ --builddirectory=_build --buildsystem=golang --with=golang
+	PATH=${PATH}:/usr/lib/go-1.18/bin dh $@ --builddirectory=_build --buildsystem=golang --with=golang
```
Comment (UmanShahzad) [inline debian/rules:3]: Pls undo the newline change

Comment (UmanShahzad) [reply debian/rules:3]:
```suggestion
	PATH=${PATH}:/usr/lib/go-1.18/bin dh $@ --builddirectory=_build --buildsystem=golang --with=golang

```

File: lib/help_detailed.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+func HelpDetailed(detailedHelp string) error {
+
```
Comment (UmanShahzad) [inline lib/help_detailed.go:11]:
```suggestion
```

File: lib/help_detailed.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	pagerCmd := os.Getenv("PAGER")
+
```
Comment (UmanShahzad) [inline lib/help_detailed.go:13]:
```suggestion
```

File: lib/help_detailed.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	if pagerCmd == "" {
+		// If PAGER is not set, use a default pager (e.g., less)
+		pagerCmd = "less"
+	}
+
+	cmd := exec.Command(pagerCmd)
+
+	// Create an io.Reader from the detailed help string
```
Comment (UmanShahzad) [inline lib/help_detailed.go:21]:
```suggestion
```

File: lib/help_detailed.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+	reader := io.Reader(strings.NewReader(detailedHelp))
+	cmd.Stdin = reader
+
```
Comment (UmanShahzad) [inline lib/help_detailed.go:24]:
```suggestion
```

File: ipinfo/cmd_default.go [hunk 19]
```diff
@@ -108,6 +110,19 @@ func cmdDefault() (err error) {
 		return nil
 	}
 
+	if fHelpDetailed {
+		// Read the string and display it using a pager
+		err = lib.HelpDetailed(DetailedHelp)
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:115]: Can we have this accept the `printHelpDefault` function as an argument and then have it called inside for the error case? Dont forget over time we'll be adding this logic everywhere and it'll just add more and more code to be copying these error cases everywhere on the outside level

File: ipinfo/cmd_default.go [hunk 20]
```diff
@@ -108,6 +110,19 @@ func cmdDefault() (err error) {
...
+
+		//if error occurs running the pager, display the default help
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error displaying detailed help: %v\n\n", err)
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:119]: We dont need to show that, just back up silently

File: ipinfo/cmd_default.go [hunk 21]
```diff
@@ -108,6 +110,19 @@ func cmdDefault() (err error) {
 		return nil
 	}
 
+	if fHelpDetailed {
+		// Read the string and display it using a pager
+		err = lib.HelpDetailed(DetailedHelp)
+
+		//if error occurs running the pager, display the default help
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:117]:
```suggestion
		// If an error occurs running the pager, display the default help.
```

File: ipinfo/cmd_default.go [hunk 22]
```diff
@@ -1,6 +1,7 @@
 package main
 
 import (
+	_ "embed"
```
Comment (UmanShahzad) [inline ipinfo/cmd_default.go:4]: Still needed?

File: lib/help_detailed.go [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+		pagerCmd = "less"
+	}
+
+	cmd := exec.Command(pagerCmd)
+	reader := io.Reader(strings.NewReader(detailedHelp))
+	cmd.Stdin = reader
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	//if an error occurs running the pager, display the default help
```
Comment (UmanShahzad) [inline lib/help_detailed.go:23]: I recommended a comment fix

File: lib/help_detailed.go [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+	if err := cmd.Run(); err != nil {
+		printHelpDefault()
+	}
+	return nil
```
Comment (UmanShahzad) [inline lib/help_detailed.go:27]:
```suggestion

	return nil
```

File: scripts/build-archive-all.sh [hunk 25]
```diff
@@ -44,4 +44,5 @@ cd ..
 rm -rf $ROOT/${CLI}/dist/usr
 mkdir -p $ROOT/${CLI}/dist/usr/local/bin
 cp $ROOT/build/${CLI}_${VSN}_linux_amd64 $ROOT/${CLI}/dist/usr/local/bin/${CLI}
-dpkg-deb -Zgzip --build ${ROOT}/${CLI}/dist build/${CLI}_${VSN}.deb
+
```
Comment (UmanShahzad) [inline scripts/build-archive-all.sh:47]: Not above, below

File: ipinfo/detailed_help.go
Comment (UmanShahzad) [inline ipinfo/detailed_help.go:1]: We dont want the detailed lookup to just be a copy/paste of the short helps of all the commands below it.

For now, let's just make it similar to the short help, and we can think about what to add collaboratively later.

File: ipinfo/detailed_help.go [hunk 26]
```diff
@@ -0,0 +1,62 @@
+package main
+
+var DetailedHelp = `Usage: ipinfo <cmd> [<opts>] [<args>]
```
Comment (UmanShahzad) [inline ipinfo/detailed_help.go:3]: We need to use the `progBase` stuff here too

File: lib/help_detailed.go [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+
+func HelpDetailed(detailedHelp string, printHelpDefault func()) error {
+	pagerCmd := os.Getenv("PAGER")
+	if pagerCmd == "" {
+		// If PAGER is not set, use a default pager (e.g., less)
+		pagerCmd = "less"
+	}
+
+	cmd := exec.Command(pagerCmd)
+	// This allows the string to be treated as an input for the command
```
Comment (UmanShahzad) [inline lib/help_detailed.go:18]:
```suggestion
```

File: ipinfo/detailed_help.go [hunk 28]
```diff
@@ -0,0 +1,65 @@
+package main
+
+import "fmt"
+
+var DetailedHelp = fmt.Sprintf(`Usage: %s <cmd> [<opts>] [<args>]
```
Comment (UmanShahzad) [inline ipinfo/detailed_help.go:5]: As we add more help messages for more subcommands, ideally it's clear which command it's for. So can we move this below the short-help in `ipinfo/cmd_default.go` and use camel case for a lower case first char?

**Review by UmanShahzad — APPROVED** [review]

**Review by UmanShahzad — APPROVED** [review]

**Pull request merged**